Add category hierarchy

okkez opened the pull request:
connect to https://github.com/bm-sms/daimon-news-misc/issues/86

okkez committed:
WIP: Introduce ancestry

okkez committed:
Cosmetic changes

okkez committed:
Add acts_as_list

okkez committed:
Use proper method name

okkez committed:
Add `:allow_nil` option to order

For acts_as_list

okkez committed:
Add test for category model

okkez committed:
Implement CategoriesController#{move_higher,move_lower}

okkez committed:
Remove link to add children

okkez committed:
Extract partials to display nested categories like tree

okkez committed:
Display roots in /editor/categories

okkez committed:
Display children of the category

okkez committed:
Add style for nested categories

okkez committed:
Add parent_id to set parent of the category

okkez committed:
Remove deleted attribute `order`

okkez committed:
Remove unneccesary test

We don't set order directly when we create category object.
User(editor) cannot set category order via form.

okkez committed:
Display root categories on navigation

okkez committed:
Extract paginate body

okkez committed:
Add seeds for nested categories

okkez committed:
Display nested categories and leaf's posts

okkez committed:
Fix style

okkez committed:
Display posts count for each categories

okkez committed:
Rename

okkez committed:
Use html options to apply select2

okkez committed:
Add help text and change `data-select2-allow-clear` to true

okkez committed:
Remove unused code

okkez committed:
Remove trailing white spaces

okkez committed:
Add named scope `leaves` to retrieve all leaves

This named scope works on PostgreSQL only.
MySQL cannot concatenate text using `||`.

See https://github.com/stefankroes/ancestry/pull/238

okkez committed:
Select category from leaves

okkez committed:
Named scope `leaves` is sorted

okkez committed:
Remove debug print

okkez committed:
Add parentheses

okkez committed:
Add integration test for category hieralchy

okkez committed:
Add parentheses

okkez committed:
Display root categories on sidebar

okkez committed:
Display category children's breadcrumb properly

okkez committed:
Increase seeds

Because I want pagination in this case.

okkez commented on app/models/category.rb:
```diff
@@ -2,8 +2,70 @@ class Category < ActiveRecord::Base
   scope :ordered, -> { order(:order) }
+  # For PostgreSQL, doesn't work on MySQL
+  # See https://github.com/stefankroes/ancestry/pull/238
+  scope :leaves, -> { joins("LEFT JOIN #{table_name} AS c ON c.#{ancestry_column} = CAST(#{table_name}.id AS text) OR c.#{ancestry_column} = #{table_name}.#{ancestry_column} || '/' || #{table_name}.id").group("#{table_name}.id").having("COUNT(c.id) = 0").order("#{table_name}.ancestry NULLS FIRST").order(:order) }
```
I don't use input from the outside, so SQL injection may not be occurred.

okkez committed:
Add named scope `parental` to Category

These categories don't have posts, so we can select these as parent
category of new category.

okkez committed:
Display categories that can become parent

okkez committed:
Simplify `Category#full_name`

okkez committed:
Display full_name to select parent category

tricknotes commented on app/assets/javascripts/editor/form.js.es6:
```diff
@@ -1,7 +1,9 @@
-window.setupPostForm = () => {
-  function applySelect2($element, options) {
+$(() => {
+  function applySelect2(selector, options) {
```
It seems `options` is currently unnecessary.
When we want to pass some options, `$element.data()` is available.

okkez replied:
I use `options` here.
https://github.com/bm-sms/daimon-news/pull/564/files#diff-bf289ed1c9275ec90f6157adf147e824R28

tricknotes replied:
Hmm, that's true.
Does `$element.find('select')` have no `data-` attributes?

okkez replied:
I can add `data-` attribute in `_credit_fields.html.haml` :yum:

tricknotes commented on app/controllers/editor/categories_controller.rb:
```diff
@@ -43,6 +43,24 @@ def destroy
   end
 
+  # move to left
+  def move_higher
```
Can you some actions except CRUD as other controller?
ref: http://jeromedalbert.com/how-dhh-organizes-his-rails-controllers/

okkez replied:
Do you mean that we should create `Editor::Categories::MoveHigherController#update` or something?

tricknotes replied:
It is close but not correct.
- 👍  Create some controller `Editor::Categories::XXXController#update`
- 👎  The controller name is not noun

The controller name should be a noun.
So the name will be `PositionController`, `OrderController` or something.

tricknotes replied:
`higher` ? `left` ?

okkez replied:
- higher is acts_as_list word
- left is ancestry word

tricknotes commented on app/models/category.rb:
```diff
@@ -2,8 +2,67 @@ class Category < ActiveRecord::Base
+  # For PostgreSQL, doesn't work on MySQL
+  # See https://github.com/stefankroes/ancestry/pull/238
+  scope :leaves, -> { joins("LEFT JOIN #{table_name} AS c ON c.#{ancestry_column} = CAST(#{table_name}.id AS text) OR c.#{ancestry_column} = #{table_name}.#{ancestry_column} || '/' || #{table_name}.id").group("#{table_name}.id").having("COUNT(c.id) = 0").order("#{table_name}.#{ancestry_column} NULLS FIRST").order(:order) }
+  scope :parental, -> { includes(:posts).where("posts.id" => nil).order("#{table_name}.#{ancestry_column} NULLS FIRST").order(:order) }
```
Can you use `quoted_table_name` instead of `table_name` ?
The former is more safe as SQL string.

tricknotes commented on app/views/editor/categories/index.html.haml:
```diff
@@ -4,24 +4,4 @@
-        %td= link_to 'Show', editor_category_url(category)
-        %td= link_to 'Edit', edit_editor_category_url(category)
-        %td= link_to 'Destroy', editor_category_url(category), method: :delete, data: { confirm: 'Are you sure?' }
+= render partial: 'table', locals: {categories: @categories}
```
``` haml
= render 'table', categories: @categories
```

tricknotes commented on app/views/editor/categories/show.html.haml:
```diff
@@ -7,9 +7,8 @@
-  %b Order
-  = @category.order
+
+= render partial: 'table', locals: {categories: @category.children.ordered}
```
``` haml
render 'table', categories: @category.children.ordered
```

okkez committed:
Reduce redandunt code

tricknotes commented on app/controllers/editor/categories_controller.rb:
```diff
@@ -43,6 +43,24 @@ def destroy
+  def move_higher
+    category = categories.find(params[:id])
+    target_category = category.higher_items.last
+    category.move_to_left_of(target_category) if target_category
```
When `target_category` is `nil`?
If it is bad case, the error should be raised?

tricknotes commented on app/controllers/editor/categories_controller.rb:
```diff
@@ -43,6 +43,24 @@ def destroy
+    target_category = category.higher_items.last
+    category.move_to_left_of(target_category) if target_category
+
+    redirect_to editor_categories_url, notice: "カテゴリが移動されました"
```
I think it is kindness to embed `category.name` into this message.

okkez committed:
Add select2 options to select element added by cocoon

okkez committed:
Remove unneccesary option

tricknotes commented on test/models/category_test.rb:
```diff
@@ -1,7 +1,49 @@
+  def setup
+    DatabaseCleaner.start
+    site = create(:site)
+    @categories = create_list(:category, 5, site: site)
```
When the some category names are the same as other one, the test cases of category's order are silently broken :<

okkez replied:
`factory :category` uses sequence for category name.
So all categories have different name each other.

On the other hand, we may need test that some categories have same name.

tricknotes replied:
> `factory :category` uses sequence for category name.

This is the current behavior and `Category` has no validation for name.
To keep this test stable, I think that to specify category name here is better way.

tricknotes commented on test/models/category_test.rb:
```diff
@@ -1,7 +1,49 @@
-  #   assert true
-  # end
+  def setup
+    DatabaseCleaner.start
```
Why is this needed?

okkez replied:
When remove `DatabaseCleaner.start`, append records for each test.

tricknotes commented on app/views/editor/categories/_form.html.haml:
```diff
@@ -2,7 +2,8 @@
   = f.text_field :slug, help: "URLの一部に使うこのカテゴリを表すわかりやすい英単語を指定します"
-  = f.number_field :order, help: "ナビゲーションバーに表示する順序を指定します"
+  - help_text = "親カテゴリを指定します。指定しない場合はトップレベルのカテゴリになります"
+  = f.collection_select :parent_id, current_site.categories.parental, :id, :full_name, {include_blank: true, help: help_text}, {class: "select2", "data-select2-allow-clear": true}
```
Except `@category` from `current_site.categories.parental` ?

okkez replied:
Oh! Fixed :bow:

okkez committed:
Extract `Editor::Categories::OrdersController`

okkez committed:
Use DatabaseCleaner properly

okkez committed:
Use `quoted_table_name` instead of `table_name`

The former is more safe as SQL string.

okkez committed:
Display categories except for the category of being updated

okkez committed:
Add custom validator to validate category

Error occurs when category has children.

okkez commented:
I've fixed against all comments and pushed changes.

tricknotes commented on app/assets/javascripts/editor/form.js.es6:
```diff
@@ -0,0 +1,30 @@
+    $element.find('.credit-order-value').val(maxOrder + 1);
+  }
+
+  applySelect2('select.select2', {});
```
The second argument is not assigned to variable in this function.

okkez committed:
Remove garbage

tricknotes commented on app/models/category.rb:
```diff
@@ -2,8 +2,78 @@ class Category < ActiveRecord::Base
+      move_to_left_of(reference_instance)
+    when :right
+      move_to_right_of(reference_instance)
+    end
```
How do you think about `else` direction?

tricknotes commented on app/validators/post_category_validator.rb:
```diff
@@ -0,0 +1,7 @@
+class PostCategoryValidator < ActiveModel::Validator
+  def validate(record)
+    if record.category.has_children?
+      record.errors[:category_id] << "子カテゴリを持つカテゴリ「#{record.category.full_name}」に記事を登録することはできません。"
```
How about using i18n for error message?
Such as https://github.com/bm-sms/nursepress/pull/272/commits/6c2a818594f35deff5f2a1f5b676f443db5c6634.

okkez replied:
In this project we don't use i18n for all messages.
If we don't need Japanese message here, I will translate to English.

okkez committed:
Add else clause for unknown direction

tricknotes commented on app/views/categories/_categories.html.haml:
```diff
@@ -0,0 +1,18 @@
+%ul.category-summary
+  - categories.each do |category|
+    - if category.has_children?
+      - count = category.children.inject(0) do |memo, child|
```
Can you use [`Enumerable#sum`](http://api.rubyonrails.org/classes/Enumerable.html#method-i-sum) to simplify calculation?

``` ruby
category.children.map {|child| child.posts.published.count }.sum
```

okkez replied:
This is the expression on the inside of deep loop.
So I want to reduce loop here.

mtsmfm replied:
> So I want to reduce loop here.

Hmm, it seems we can't reduce loop at any rate.

In this case, `sum` is more explanatory than `inject` I think.

---

FYI: `sum` accepts block:

``` ruby
count = category.children.sum {|child| child.posts.published.count }
```

okkez replied:
https://github.com/rails/rails/blob/a678c47f6d31977e2a91cc198f115c803f0fc8e1/activesupport/lib/active_support/core_ext/enumerable.rb#L20

`sum` accepts block, but calls `map` and `inject`.

In my case, calls `inject` only and loop children at once.

In tricknotes case,

``` ruby
category.children.map {|child| child.posts.published.count }.sum
```

This invokes `map` and `inject` (in `sum`), loop children and array (has same size with children).
loop twice.

In mtsmfm case,

``` ruby
count = category.children.sum {|child| child.posts.published.count }
```

This invokes `map` in `sum` and call `sum` recursively and call `inject`.
After all, loop twice.

tricknotes commented on config/routes.rb:
```diff
@@ -17,7 +17,9 @@
     resources :links
-    resources :categories
+    resources :categories do
+      resources :orders, module: :categories, only: %i(update), param: :direction, constraints: {direction: /left|right/}
```
This route generates the following URL.

```
PUT /editor/categories/:category_id/orders/:direction
```

I think this is not RESTFul because `:direction` is not a resource.

Instead, how about the following URL?

```
POST /editor/categories/:category_id/order?move_to=:direction
```

okkez committed:
Change `#update` to `#create`

okkez committed:
Singularlize URL

Before: POST /editor/categories/:category_id/orders
After: POST /editor/categories/:category_id/order

tricknotes commented on app/models/category.rb:
```diff
@@ -2,8 +2,80 @@ class Category < ActiveRecord::Base
+  def move_to_child_of(reference_instance)
+    transaction do
+      remove_from_list
+      update_attributes!(parent: reference_instance)
```
Can you use `update!` (it is the alias of `update_attributes!`) for consistency?

```
$ git grep "update\!" | wc
       9      26     658
$ git grep "update_attributes\!" | wc
       0       0       0
```

tricknotes commented on app/models/category.rb:
```diff
@@ -2,8 +2,80 @@ class Category < ActiveRecord::Base
+
   validates :slug, format: /\A\w+\z/, uniqueness: {scope: :site_id}
-  validates :order, numericality: :only_integer
+  validates :order, numericality: {only_integer: true, allow_nil: true}
```
When the order becomes `nil`?

okkez replied:
When `remove_from_list`, this is `acts_as_list`'s method.

tricknotes replied:
confirmed

okkez committed:
Use `update!` instead of `update_attributes!`

For consistency.

mtsmfm commented on app/controllers/editor/categories/orders_controller.rb:
```diff
@@ -0,0 +1,26 @@
+class Editor::Categories::OrdersController < Editor::ApplicationController
+  def create
```
`update` ?

okkez committed:
Use `#update` instead of `#create`

`#update` is prefer in this case.

okkez committed:
Use `Enumerable#sum` instead of `Array#inject`

kou commented on app/views/categories/_categories.html.haml:
```diff
@@ -0,0 +1,18 @@
+%ul.category-summary
+  - categories.each do |category|
+    - if category.has_children?
+      - count = category.children.sum do |memo, child|
```
The first block argument `memo` is needed?

okkez commented:
Rebased

tricknotes commented on app/assets/javascripts/editor/form.js.es6:
```diff
@@ -0,0 +1,30 @@
+
+    orders.push(0); // Minimum value
+
+    let maxOrder = Math.max(...orders);
```
We can simplify 🙇 💦 

``` js
Math.max(0, ...orders);
```

okkez replied:
Could you fix this after merge this PR?

tricknotes replied:
nod

tricknotes commented on app/controllers/editor/categories/orders_controller.rb:
```diff
@@ -0,0 +1,26 @@
+    current_site.categories
+  end
+
+  def category_params
```
Maybe is this unused?

tricknotes commented on app/views/editor/categories/_category.html.haml:
```diff
@@ -0,0 +1,11 @@
+  %td= link_to 'Show', editor_category_url(category)
+  %td= link_to 'Edit', edit_editor_category_url(category)
+  %td= link_to 'Destroy', editor_category_url(category), method: :delete, data: { confirm: 'Are you sure?' }
+  %td= link_to_unless category.first?, '▲', editor_category_order_url(category_id: category, move_to: "left", target: category.higher_item&.id), method: :patch
```
Should this HTTP method be the `PUT`?

okkez replied:
I'm not sure.

tricknotes replied:
> PUT to a URL creates/replaces the resource in its entirety at the client defined URL.
> PATCH to a URL updates part of the resource at that client defined URL.

ref: http://stackoverflow.com/questions/630453/put-vs-post-in-rest

In this case, these request parameters are the part of resource or itself?

okkez replied:
I got it :clap:

tricknotes commented on app/models/category.rb:
```diff
@@ -2,8 +2,80 @@ class Category < ActiveRecord::Base
+  end
+
+  def move_to(reference_instance, direction:)
+    case direction.to_sym
```
Are there any reason using `to_sym`?
I think this should be a `String` because `direction` comes from user input.

okkez replied:
Just normalized.

tricknotes replied:
nod

tricknotes commented on db/seeds/development/posts.seeds.rb:
```diff
@@ -70,20 +72,53 @@
   end
+
+  25.times do |i|
+    i += site1.posts.maximum(:id)
```
We can use `public_id` instead `id`.
The latter is not published anywhere.

tricknotes commented on test/models/category_test.rb:
```diff
@@ -1,7 +1,44 @@
+    assert_equal(expected, Category.ordered.pluck(:name))
+  end
+
+  def test_move_to_left_of_2
```
What is the difference between 1 and 2?
I'm happy if it is described at test case.

okkez committed:
Remove unused method

okkez committed:
Use public_id instead of id

Because `public_id > id` here.

okkez committed:
More descriptive test method

tricknotes committed:
Fix to use `Enumerable#sum` correctly

tricknotes committed:
[Refactor] Move logic to model

okkez committed:
Use `PUT` instead of `PATCH`

> PUT to a URL creates/replaces the resource in its entirety at the client defined URL.
> PATCH to a URL updates part of the resource at that client defined URL.

ref: http://stackoverflow.com/questions/630453/put-vs-post-in-rest

okkez commented:
Rebased

tricknotes commented on app/models/category.rb:
```diff
@@ -2,8 +2,90 @@ class Category < ActiveRecord::Base
   scope :ordered, -> { order(:order) }
+  # For PostgreSQL, doesn't work on MySQL
+  # See https://github.com/stefankroes/ancestry/pull/238
+  scope :leaves, -> { joins("LEFT JOIN #{quoted_table_name} AS c ON c.#{ancestry_column} = CAST(#{quoted_table_name}.id AS text) OR c.#{ancestry_column} = #{quoted_table_name}.#{ancestry_column} || '/' || #{quoted_table_name}.id").group("#{quoted_table_name}.id").having("COUNT(c.id) = 0").order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST").order(:order) }
```
Can you specify `order` only once?

``` diff
- order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST").order(:order)
+ order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST", :order)
```

tricknotes commented on app/models/category.rb:
```diff
@@ -2,8 +2,90 @@ class Category < ActiveRecord::Base
+  # For PostgreSQL, doesn't work on MySQL
+  # See https://github.com/stefankroes/ancestry/pull/238
+  scope :leaves, -> { joins("LEFT JOIN #{quoted_table_name} AS c ON c.#{ancestry_column} = CAST(#{quoted_table_name}.id AS text) OR c.#{ancestry_column} = #{quoted_table_name}.#{ancestry_column} || '/' || #{quoted_table_name}.id").group("#{quoted_table_name}.id").having("COUNT(c.id) = 0").order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST").order(:order) }
+  scope :parental, ->(category = nil) { includes(:posts).where("posts.id" => nil).where.not(id: category&.id).order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST").order(:order) }
```
`order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST").order(:order)` is used twice.
Can you extract it as scope?

okkez replied:
Good catch. But I could not find out good name for this ...
I want to extract this when I write same code 3 times.

tricknotes commented on test/models/category_test.rb:
```diff
@@ -1,7 +1,44 @@
-  # end
+  def setup
+    site = create(:site)
+    @categories = create_list(:category, 5, site: site)
```
Can you specify each category to individual name?
Currently `Category#name` depends on data initialization.

tricknotes commented on test/test_helper.rb:
```diff
@@ -13,4 +13,5 @@
 class ActiveSupport::TestCase
   # Add more helper methods to be used by all tests here...
   include FactoryGirl::Syntax::Methods
+  self.use_transactional_fixtures = false
```
👍

tricknotes commented on test/models/category_test.rb:
```diff
@@ -1,7 +1,44 @@
+  def test_move_to_left_of
+    @categories[4].move_to_left_of(@categories[3])
+    expected = @categories.map(&:name).values_at(0, 1, 2, 4, 3)
+    assert_equal(expected, Category.ordered.pluck(:name))
```
This test case depends on `Category.ordered`.
I think this test case is a unit test, so it shouldn't extra logic.

okkez replied:
I want to test ordered categories after move categories.

---

入れ替えたあとのカテゴリの並びをテストしたい、という説明をしてご納得いただいた。

tricknotes replied:
👍

tricknotes commented on test/integration/editor_test.rb:
```diff
@@ -307,25 +306,122 @@ class EditorTest < ActionDispatch::IntegrationTest
+      click_on("New Category")
+
+      fill_in("Name",        with: "JRuby")
+      fill_in("Description", with: "JRuby is a one of Ruby implementation.")
```
"JRuby is an implementation of Ruby" ?

okkez committed:
Simplify

okkez committed:
test: Fix description

okkez committed:
Set category name when initialize test data

Category test will not be broken when we change logic to generate
category name in factory.

mtsmfm commented:
:shipit:

mtsmfm commented:
I released migration.
Could you rebase this pr?

okkez committed:
Reduce ABC size

mtsmfm commented:
:shipit:
